crates/vm2/src/world_diff.rs

```diff
@@ -369,105 +375,195 @@ const COLD_WRITE_AFTER_WARM_READ_REFUND: u32 = STORAGE_ACCESS_COLD_READ_COST;
-                        }
-                    ))
-                    .collect()
-            );
-
-            let checkpoint2 = world_diff.snapshot();
-            for (key, value) in &second_changes {
-                world_diff.write_storage(&mut NoWorld, &mut (), key.0, key.1, *value);
-            }
-            assert_eq!(
-                world_diff
-                    .get_storage_changes_after(&checkpoint2)
-                    .collect::<BTreeMap<_, _>>(),
-                second_changes
-                    .iter()
-                    .map(|(key, value)| (
-                        *key,
-                        StorageChange {
-                            before: first_changes.get(key).or(initial_values.get(key)).copied(),
-                            after: *value,
-                            is_initial: !initial_values.contains_key(key),
-                        }
-                    ))
-                    .collect()
-            );
-
-            let mut combined = first_changes
-                .into_iter()
-                .filter_map(|(key, value)| {
-                    let initial = initial_values.get(&key).copied();
-                    (initial.unwrap_or_default() != value).then_some((key, (initial, value)))
-                })
-                .collect::<BTreeMap<_, _>>();
-            for (key, value) in second_changes {
-                let initial = initial_values.get(&key).copied();
-                if initial.unwrap_or_default() == value {
-                    combined.remove(&key);
-                } else {
-                    combined.insert(key, (initial, value));
-                }
-            }
+            test_storage_changes(&initial_values, first_changes, second_changes);
+        }
 
-            assert_eq!(combined, world_diff.get_storage_changes().collect());
+        #[test]
+        fn storage_changes_work_with_constrained_changes(
+            initial_values in constrained_initial_storage(),
+            first_changes in constrained_storage_changes(),
+            second_changes in constrained_storage_changes(),
+        ) {
+            test_storage_changes(&initial_values, first_changes, second_changes);
         }
     }
 
+    /// Max items in generated initial storage / changes.
+    const MAX_ITEMS: usize = 5;
+    /// Bit mask for bytes in constrained `U256` / `H160` values.
+    const BIT_MASK: u8 = 0b_1111;
+
+    fn arbitrary_initial_storage() -> impl Strategy<Value = BTreeMap<(H160, U256), StorageSlot>> {
+        btree_map(
+            any::<([u8; 20], [u8; 32])>()
+                .prop_map(|(contract, key)| (H160::from(contract), U256::from(key))),
+            any::<([u8; 32], bool)>().prop_map(|(value, is_write_initial)| StorageSlot {
+                value: U256::from(value),
+                is_write_initial,
+            }),
+            0..=MAX_ITEMS,
+        )
+    }
+
+    fn constrained_initial_storage() -> impl Strategy<Value = BTreeMap<(H160, U256), StorageSlot>> {
```

Comment:
The unconstrained strategy fails to catch at least some simple bugs, e.g. setting `is_initial: initial.value.is_zero()` in `get_storage_changes_after()`.